Add unboxedCallRedispatch

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35714 Add temporary impl_UNBOXED syntax sugar for unboxed-only defs.
* #35706 Add DispatchKey impl overload; remove use of torch::dispatch
* **#35476 Add unboxedCallRedispatch**

A few things:
- Add new callUnboxedRedispatch function which can be used to do a
  redispatch when you don't want to add a type id to the excluded
  set.  This will recompute the dispatch key but ignore everything
  including and before the currentDispatchKey
- Add FULL_AFTER constructor to DispatchKeySet; used to implement
  redispatch.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20680518](https://our.internmc.facebook.com/intern/diff/D20680518)

ezyang committed:
Add BackendGeneric dispatch key and unboxedCallRedispatch

A few things:
- Add a new BackendGeneric dispatch key, intended to be a place where
  we can put code which is generic across backends.  No operators
  are modified to use it but there is a simple use case test
  in op_registration_test.cpp.  It's enabled by default (like
  Variable) but fallthrough by default (like BackendSelect)
- Add new callUnboxedRedispatch function which can be used to do a
  redispatch when you don't want to add a type id to the excluded
  set.  This will recompute the dispatch key but ignore everything
  including and before the currentDispatchKey
- Add FULL_AFTER constructor to DispatchKeySet; used to implement
  redispatch.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

smessmer reviewed (changes requested):
Why do we introduce a new `BackendGeneric` instead of just reusing `BackendSelect` and potentially renaming it?

ezyang committed:
Update on "Add BackendGeneric dispatch key and unboxedCallRedispatch"


A few things:
- Add a new BackendGeneric dispatch key, intended to be a place where
  we can put code which is generic across backends.  No operators
  are modified to use it but there is a simple use case test
  in op_registration_test.cpp.  It's enabled by default (like
  Variable) but fallthrough by default (like BackendSelect)
- Add new callUnboxedRedispatch function which can be used to do a
  redispatch when you don't want to add a type id to the excluded
  set.  This will recompute the dispatch key but ignore everything
  including and before the currentDispatchKey
- Add FULL_AFTER constructor to DispatchKeySet; used to implement
  redispatch.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20680518](https://our.internmc.facebook.com/intern/diff/D20680518)

[ghstack-poisoned]

ezyang commented:
> Why do we introduce a new BackendGeneric instead of just reusing BackendSelect and potentially renaming it?

Because... they have different semantic meanings?

ezyang committed:
Update on "Add BackendGeneric dispatch key and unboxedCallRedispatch"


A few things:
- Add a new BackendGeneric dispatch key, intended to be a place where
  we can put code which is generic across backends.  No operators
  are modified to use it but there is a simple use case test
  in op_registration_test.cpp.  It's enabled by default (like
  Variable) but fallthrough by default (like BackendSelect)
- Add new callUnboxedRedispatch function which can be used to do a
  redispatch when you don't want to add a type id to the excluded
  set.  This will recompute the dispatch key but ignore everything
  including and before the currentDispatchKey
- Add FULL_AFTER constructor to DispatchKeySet; used to implement
  redispatch.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20680518](https://our.internmc.facebook.com/intern/diff/D20680518)

[ghstack-poisoned]

ezyang committed:
Update on "Add BackendGeneric dispatch key and unboxedCallRedispatch"


A few things:
- Add a new BackendGeneric dispatch key, intended to be a place where
  we can put code which is generic across backends.  No operators
  are modified to use it but there is a simple use case test
  in op_registration_test.cpp.  It's enabled by default (like
  Variable) but fallthrough by default (like BackendSelect)
- Add new callUnboxedRedispatch function which can be used to do a
  redispatch when you don't want to add a type id to the excluded
  set.  This will recompute the dispatch key but ignore everything
  including and before the currentDispatchKey
- Add FULL_AFTER constructor to DispatchKeySet; used to implement
  redispatch.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20680518](https://our.internmc.facebook.com/intern/diff/D20680518)

[ghstack-poisoned]

smessmer commented:
> > Why do we introduce a new BackendGeneric instead of just reusing BackendSelect and potentially renaming it?
> 
> Because... they have different semantic meanings?

They're different use cases but they're the same mechanism of "do some work before calling into the actual kernel". I was thinking that maybe we can unify them.

ezyang committed:
Update on "Add BackendGeneric dispatch key and unboxedCallRedispatch"


A few things:
- Add a new BackendGeneric dispatch key, intended to be a place where
  we can put code which is generic across backends.  No operators
  are modified to use it but there is a simple use case test
  in op_registration_test.cpp.  It's enabled by default (like
  Variable) but fallthrough by default (like BackendSelect)
- Add new callUnboxedRedispatch function which can be used to do a
  redispatch when you don't want to add a type id to the excluded
  set.  This will recompute the dispatch key but ignore everything
  including and before the currentDispatchKey
- Add FULL_AFTER constructor to DispatchKeySet; used to implement
  redispatch.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20680518](https://our.internmc.facebook.com/intern/diff/D20680518)

[ghstack-poisoned]

ezyang committed:
Update on "Add BackendGeneric dispatch key and unboxedCallRedispatch"


A few things:
- Add a new BackendGeneric dispatch key, intended to be a place where
  we can put code which is generic across backends.  No operators
  are modified to use it but there is a simple use case test
  in op_registration_test.cpp.  It's enabled by default (like
  Variable) but fallthrough by default (like BackendSelect)
- Add new callUnboxedRedispatch function which can be used to do a
  redispatch when you don't want to add a type id to the excluded
  set.  This will recompute the dispatch key but ignore everything
  including and before the currentDispatchKey
- Add FULL_AFTER constructor to DispatchKeySet; used to implement
  redispatch.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20680518](https://our.internmc.facebook.com/intern/diff/D20680518)

[ghstack-poisoned]

ezyang committed:
Update on "Add unboxedCallRedispatch"


A few things:
- Add new callUnboxedRedispatch function which can be used to do a
  redispatch when you don't want to add a type id to the excluded
  set.  This will recompute the dispatch key but ignore everything
  including and before the currentDispatchKey
- Add FULL_AFTER constructor to DispatchKeySet; used to implement
  redispatch.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20680518](https://our.internmc.facebook.com/intern/diff/D20680518)

[ghstack-poisoned]

ezyang commented:
@smessmer I updated this diff to not include BackendGeneric anymore.

ezyang committed:
Update on "Add unboxedCallRedispatch"


A few things:
- Add new callUnboxedRedispatch function which can be used to do a
  redispatch when you don't want to add a type id to the excluded
  set.  This will recompute the dispatch key but ignore everything
  including and before the currentDispatchKey
- Add FULL_AFTER constructor to DispatchKeySet; used to implement
  redispatch.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20680518](https://our.internmc.facebook.com/intern/diff/D20680518)

[ghstack-poisoned]

smessmer commented on c10/core/DispatchKeySet.h:
```diff
@@ -34,6 +34,7 @@ namespace c10 {
 class DispatchKeySet final {
 public:
   enum Full { FULL };
```
Thinking about it, I probably would have put those into static factory functions instead of constructors with an enum argument, i.e.

    static DispatchKeySet DispatchKeySet::full();
    static DispatchKeySet DispatchKeySet::raw();
    static DispatchKeySet DispatchKeySet::full_after(...);

Seems more standard. But not a strong opinion if you wanna keep it as is.

ezyang replied:
I'm in the habit of doing constructors because it makes field initialization syntax better (doesn't matter here as we have copy/move constructors but when you don't have those it makes a ton of difference.)

ezyang committed:
Update on "Add unboxedCallRedispatch"


A few things:
- Add new callUnboxedRedispatch function which can be used to do a
  redispatch when you don't want to add a type id to the excluded
  set.  This will recompute the dispatch key but ignore everything
  including and before the currentDispatchKey
- Add FULL_AFTER constructor to DispatchKeySet; used to implement
  redispatch.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20680518](https://our.internmc.facebook.com/intern/diff/D20680518)

[ghstack-poisoned]
